File: exo/python/src/device_lifecycle.py
from __future__ import annotations

import argparse
import time

from qnbot_sdk import Health, Sdk, SerialConnection
from qnbot_sdk.exo import ExoConfig


def create_sdk(first_port: str, second_port: str) -> Sdk:
    return Sdk(
        devices=(
            ExoConfig(name="first", connection=SerialConnection(port=first_port)),
            ExoConfig(name="second", connection=SerialConnection(port=second_port)),
        )
    )


def print_health(label: str, health: Health) -> None:
    print(
        f"{label} health ok={health.ok} warnings={health.warning_count} "
        f"errors={health.error_count}"
    )


def main() -> None:
    arguments = argparse.ArgumentParser(description="Manage two Exo devices apart")
    arguments.add_argument("--first-port", required=True)
    arguments.add_argument("--second-port", required=True)
    arguments.add_argument("--seconds", type=float, default=3.0)
    options = arguments.parse_args()

    sdk = create_sdk(options.first_port, options.second_port)
    exo = sdk.exo()
    first = exo.device("first")
    second = exo.device("second")
    first.start()
    second.start()
    print_health("first", first.health())
    print_health("second", second.health())
    time.sleep(options.seconds)

    second.stop()
    print("second device stopped; first remains active", flush=True)
    print_health("first", first.health())
    latest = first.telemetry().latest()
    if latest is not None:
        print(
            f"first telemetry sequence={latest.sequence} "
            f"left_arm={latest.value.payload.left_arm.encoder_counts}"
        )

    first.stop()
    first.close()
    second.close()


if __name__ == "__main__":
    main()

File: exo/cpp/src/device_lifecycle.cpp
#include "exo_example_support.hpp"

#include <chrono>
#include <cstdlib>
#include <exception>
#include <iostream>
#include <string>
#include <thread>
#include <utility>

int main(int argc, char** argv) {
    try {
        std::string first_port;
        std::string second_port;
        double seconds = 3.0;
        for (int index = 1; index < argc; ++index) {
            const std::string argument = argv[index];
            if (argument == "--first-port") {
                first_port =
                    example::require_value(argc, argv, index, argument);
            } else if (argument == "--second-port") {
                second_port =
                    example::require_value(argc, argv, index, argument);
            } else if (argument == "--seconds") {
                seconds = example::parse_seconds(
                    example::require_value(argc, argv, index, argument));
            } else {
                throw std::invalid_argument("unknown argument: " + argument);
            }
        }
        if (first_port.empty() || second_port.empty()) {
            throw std::invalid_argument(
                "--first-port and --second-port are required");
        }

        qnbot::ExoConfig first_config;
        first_config.connection.port = first_port;
        first_config.name = "first";
        qnbot::ExoConfig second_config;
        second_config.connection.port = second_port;
        second_config.name = "second";
        qnbot::SdkConfig config;
        config.devices = {first_config, second_config};

        qnbot::Sdk sdk(std::move(config));
        auto exo = sdk.exo();
        auto first = exo.device("first");
        auto second = exo.device("second");
        first.start();
        second.start();
        example::print_health("first", first.health());
        example::print_health("second", second.health());
        std::this_thread::sleep_for(std::chrono::duration<double>(seconds));

        second.stop();
        std::cout << "second device stopped; first remains active" << std::endl;
        example::print_health("first", first.health());
        const auto latest = first.telemetry().latest();
        if (latest) {
            std::cout << "first telemetry sequence=" << latest->sequence
                      << " left_arm=";
            example::print_encoder_counts(
                latest->value.payload.left_arm.encoder_counts);
            std::cout << '\n';
        }

        first.stop();
        first.close();
        second.close();
        return EXIT_SUCCESS;
    } catch (const std::exception& error) {
        return example::report_error(error);
    }
}
